### PR TITLE
[Experimental] Product filters > Attributes loading state in editor

### LINKED ISSUE #50150: Product Filters Redesign: Attribute Filter: Loading State
It should pulse, slowly changing opacity from 5% to 8% and back.

image

Figma: 3td38XZM4P9WhWg618nXjt-fi-64_167456

### PR DESCRIPTION
**Pull request by roykho**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/50150

This PR adds a loading state when attributes' terms are loading. This is so that during loading (when network is slow), the user can visually see a loading representation of the state. I have modeled this the same as the rating block to be consistent.

Google Chrome - Product Filters (Experimental) ‹ Template part ‹ woo-blocks-dev ‹ Editor — WordPress 2024-09-04 at 7 22 35 AM

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Test by developers only**

**Prereq:**

Ensure you have the WC Beta Tester Plugin installed and go to Tools > WCA Test Helper > Features. Toggle on the experimental-blocks option.

1. Go to the editor and navigate to Patterns->General template parts->Product Filters (Experimental).
2. Using the browser's console, throttle your network to a slower setting.
3. Refresh the page and you should see the initial loading state of the attributes before the terms are populated.
4. Click on the filter options, change the attribute to another term and ensure you also see the pulsing loading state.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 
Product filters: add loading state for attributes filter
</details>

### CODE REVIEW

**Review by dinhtungdu — CHANGES_REQUESTED** [review]
Thanks for working on the PR. I can see the bars were loading and pulsing in my test. However, there is an issue with rerender that needs to be addressed before merging this PR which I shared in the review comment.

**Comment by roykho** [comment]
Hi @dinhtungdu I am not seeing your review comment?

**Review by thealexandrelara — APPROVED** [review]
Thank you for working on this @roykho! Code looks good, and everything passed in my tests so I'm approving this PR.

## Test summary:

✅ Scenario 1: A merchant with a slow connection visits the Patterns list page
- Expected Outcome: The Attribute filter blocks displayed in the pattern/template part preview should appear in a loading state until the content for those blocks is fully loaded.
- Details: The Product Filters preview displays the loading indicator for the attribute filters while the content is loading.

---


✅ Scenario 2: A merchant with a slow connection visits the Product Filters (Experimental) template part.
- Expected Outcome: The Attribute filter block displays the Heading block, and the Filter Options remain in a loading state until the content is fully loaded.
- Details: When the content of the Filter Options for the Attribute filter blocks is not yet loaded, the Filter Options display the loading indicator.

---

✅ Scenario 3: A merchant with a slow connection visits the Product Filters (Experimental) template part in a store without any registered attribute.
- Expected Outcome: A message stating that no attributes have been created for the store is displayed instead of the loading indicator.
- Details: When no attributes are registered for the store, the message is correctly displayed, prompting users to create one.

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/edit.tsx [hunk 1]
```diff
@@ -144,7 +149,31 @@ const Edit = ( props: EditProps ) => {
...
+										Math.floor(
+											Math.random() * ( 100 - 25 )
+										) + '%',
+								} }
+							>
+								&nbsp;
+							</li>
+						);
+					} ) }
+				</ul>
```
Comment (dinhtungdu) [inline plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/edit.tsx]: @roykho I'm sorry, not sure why this wasn't submitted.

Without using `useMemo`, the bars' width is changed between rerenders.

https://github.com/user-attachments/assets/6f15a9f6-5290-4fc8-91a4-61084f459b6e

**Comment by roykho** [comment]
@dinhtungdu thanks for the review. In my test, I didn't click on the individual blocks to make it re-render so I missed this. Good catch. Fixed in https://github.com/woocommerce/woocommerce/pull/51151/commits/bb489cb7f944a9e8ae2834675c81add2b696aaf5

**Review by dinhtungdu — APPROVED** [review]
Thanks for the quick update! I tested it again and this PR is working great in my test.

**Pull request merged**